rfc: console sign in process (#5445)
winglang/wing: docs/contributing/999-rfcs/2024-01-15-console-sign-in.md

---
title: "Wing Console Sign In"
description: Requirements for Wing Console Sign In process
---

# Wing Console Sign In

- **Author(s)**: @ainvoner
- **Submission Date**: 2024-01-15
- **Stage**: Proposal
- **Stage Date**: 2024-01-15

This document describes the Wing Console Sign In process and rationale.

## Background

> As we seek to find **Sales and Marketing Opportunities**, it has become increasingly important to establish a more direct and meaningful connection with our users.

Reviewer: ```suggestion
As we seek to be able to understand how wing is used, it has become increasingly important to establish a more direct and meaningful connection with our users.
```

One of the ways to achieve this is by implementing a sign-in flow within the Wing Console application.

## Requirements

> New users of Wing Console will be required to sign in with their GitHub account before having access to the Wing Console application.

Reviewer 2: The more enterprisier the audience, the more likely it'll be that people might not have a Github account. Any chance to track this somehow?

Reviewer: How hard is it to add additional providers? If not too difficult then maybe we should.

Alternatively, we can add a "I don't have a GitHub account" link which will skip sign in and we can track analytics.

Author: if we will allow to skip sign - no one will sign in.
I think, for initial phase, support only GitHub (we can add additional sign in providers later).

Reviewer 3: +1

Reviewer 4: We would need to do some refactor in Wing Cloud to allow other providers. It'd take some time right now.

> Alternatively, we can add a "I don't have a GitHub account" link which will skip sign in and we can track analytics.

That's what we do now, right? Just anonymous analytics.

Another requirement is to have only one user entity in Wing. This means that a user in Wing Console is also a user in Wing Cloud and vice versa.

## User Experience

### New Users

> New users that will `wing it` will see a sign in dialog blocking the Wing Console UI. The rationale behind this decision is that we want to show the user a teaser of Wing Console while asking them to sign in.

Reviewer: What's the relationship to the EULA?

Maybe the license agreement is a button from the sign up screen:

> By using the Wing Console you agree to these [terms and conditions]() 

or something like this.

Reviewer 4: Who knows about this legal stuff?

> New users that will `wing it` will see a sign in dialog blocking the Wing Console UI. The rationale behind this decision is that we want to show the user a teaser of Wing Console while asking them to sign in. 

Reviewer 2: It should be made clear, that everything is still running locally. The user might get confused and believe that with the sign up, stuff will be running remotely. Questions like "will I expose my highly confidential code by doing this" should be addressed.

Once users will be signed in with their GitHub account they will be able to use Wing Console.

### Singed In Users

> Signed in users to Wing Console will not need to sign in again. They will be able to use Wing Console as usual.

Reviewer: Worth mentioning that the user is per machine so any time the console is opened, if there was already a logged in user, it will be used.

Reviewer 4: Yes, the relationship of the analytics ID and the Wing Cloud user will only happen once. We won't require this process again anymore.

### Offline

> If the user is offline (no internet connection) they will be able to use Wing Console as usual without the need to sign in.

Reviewer: I am assuming this is an issue only for the first sign up, right?

Reviewer 4: Exactly.

### Support of multiple users on the same machine

There would be no support for multiple users on the same machine. 

## User identity for Analytics  

Events generated by signed in users will include the user's GitHub account email as an integral part of their user identity.
In contrast, Wing CLI users are initially anonymous, and all analytics events originating from the CLI are sent anonymously with a unique identifier.

> However, when a user signs in to Wing Console, the unique identifier associated with CLI analytics events will be associated with the signed in user's identity, allowing us to link CLI analytics data to specific user accounts

Reviewer 2: What about logging out locally?

## Technical details

Since we would like to have a single user entity in Wing, the process of signing in to Wing Console will use the same authentication mechanism as Wing Cloud.
In fact, the user will actually be singing in to Wing Cloud and be redirected back to Wing Console without noticing.

> In order to make sure a sign in user won't need to sign in again, we will save the uses' authentication token in user's machine.

Reviewer 2: ```suggestion
In order to make sure a sign in user won't need to sign in again, we will save the uses' authentication token in user's machine.

## Next Steps

Add value for the user in exchange for signin up. Things which could be considered:

- Deploy to wing.cloud directly from the console. This could upload the working directory to a Bucket from which wing.cloud could deploy. The deployment status and the url would be shown in the local console.
- Offer a solution similar to ngrok, to make the locally running instance available for remote services (e.g. webhooks) or users (e.g. demo purposes)
```

One of the ways to achieve this is by implementing a sign-in flow within the Wing Console application.

## Requirements

New users of Wing Console will be required to sign in with their GitHub account before having access to the Wing Console application.

Another requirement is to have only one user entity in Wing. This means that a user in Wing Console is also a user in Wing Cloud and vice versa.

## User Experience

### New Users

New users that will `wing it` will see a sign in dialog blocking the Wing Console UI. The rationale behind this decision is that we want to show the user a teaser of Wing Console while asking them to sign in.

Once users will be signed in with their GitHub account they will be able to use Wing Console.

### EULA (End User License Agreement) for new users

The license agreement should be added to the bottom of the sign in screen:

"By using the Wing Console you agree to these [terms and conditions](#)"

### Singed In Users

Signed in users to Wing Console will not need to sign in again. They will be able to use Wing Console as usual.

### Offline

> If the user is offline (no internet connection) they will be able to use Wing Console as usual without the need to sign in.

Reviewer 3: Assuming they already signed in right?

Reviewer 4: If they are offline, they won't be able to sign in anyway so we should allow them to just use the Console directly.
